### PR TITLE
Add labels

### PR DESCRIPTION
**Pull request by Gregory-Pereira**
This pull requests implements a configuration labelmap to map `alertmanager` label values to GitHub issue label values.

Addresses: https://github.com/operate-first/support/issues/744

### CODE REVIEW

**Comment by Gregory-Pereira** [comment]
@HumairAK please help

**Comment by Gregory-Pereira** [comment]
Go does not support function with variable-length arrays, must find a way to initialized the function with structured string array size for the NewReciever funciton

File: alerts/template_test.go [hunk 1]
```diff
@@ -20,11 +20,22 @@ import (
 	"html/template"
 	"strings"
 	"testing"
+	"reflect"
 
 	"github.com/prometheus/alertmanager/notify/webhook"
 	amtmpl "github.com/prometheus/alertmanager/template"
 )
 
+func containsArray(arr []string, str string) (bool) {
```
Comment (HumairAK) [inline alerts/template_test.go]: This doesn't look like it's used anywhere

File: alerts/template_test.go [hunk 2]
```diff
@@ -93,3 +153,115 @@ func TestFormatTitleSimple(t *testing.T) {
...
 }
+
+func TestFormatLabelsSimple(t *testing.T) {
+	msg := webhook.Message{
+		Data: &amtmpl.Data{
+			Status: "firing",
+			Alerts: []amtmpl.Alert{
+				{
+					Annotations: amtmpl.KV{"env": "prod", "svc": "foo"},
+					Labels: amtmpl.KV{"cluster": "rick", "namespace": "opf-ci-pipelines"},
```
Comment (HumairAK) [inline alerts/template_test.go]: can you switch these to use none operate-first related labels? anything generic works (same as above)

File: alerts/template_test.go [hunk 3]
```diff
@@ -93,3 +153,115 @@ func TestFormatTitleSimple(t *testing.T) {
...
+				},
+				{
+					Annotations: amtmpl.KV{"env": "stage", "svc": "foo"},
+					Labels: amtmpl.KV{"cluster": "smaug", "namespace": "openshift-monitoring"},
```
Comment (HumairAK) [inline alerts/template_test.go]: can you switch these to use none operate-first related labels? anything generic works (same as above)

File: alerts/template_test.go [hunk 4]
```diff
@@ -46,34 +57,83 @@ func TestFormatTitleSimple(t *testing.T) {
 			Alerts: []amtmpl.Alert{
 				{
 					Annotations: amtmpl.KV{"env": "prod", "svc": "foo"},
+					Labels: amtmpl.KV{"cluster": "rick", "namespace": "opf-ci-pipelines"},
```
Comment (HumairAK) [inline alerts/template_test.go]: can you switch these to use none operate-first related labels? anything generic works, like: 
```
Labels: amtmpl.KV{"testkey1": "testval1", "testkey2": "testval2"},
```

File: alerts/template_test.go [hunk 5]
```diff
@@ -46,34 +57,83 @@ func TestFormatTitleSimple(t *testing.T) {
...
 				},
 				{
 					Annotations: amtmpl.KV{"env": "stage", "svc": "foo"},
+					Labels: amtmpl.KV{"cluster": "smaug", "namespace": "openshift-monitoring"},
```
Comment (HumairAK) [inline alerts/template_test.go]: can you switch these to use none operate-first related labels? anything generic works (same as above)

File: alerts/template_test.go [hunk 6]
```diff
@@ -93,3 +153,115 @@ func TestFormatTitleSimple(t *testing.T) {
...
+	expectOutputTest3[0] = "firing"
+	var expectOutputTest4 = make([]string, 4)
+	expectOutputTest4[0] = "rick"
+	expectOutputTest4[1] = "opf-ci-pipelines"
+	expectOutputTest4[2] = "smaug"
+	expectOutputTest4[3] = "openshift-monitoring"
+	var expectOutputTest5 = make([]string, 2)
+	expectOutputTest5[0] = "prod"
+	expectOutputTest5[1] = "stage"
+	var expectOutputTest6 = make([]string, 0)
```
Comment (HumairAK) [inline alerts/template_test.go]: I would get rid of these assignments and include them directly in the struct creations, e.g.: 

 ```
 {
	testName: "test status directly", 
	labelsTmpl: "{{.Status}}", 
	expectErrTxt: "", 
	expectOutput: []string{"firing"},   // get rid of expectOutputTest3 entirely
}
 ```

Comment (HumairAK) [reply alerts/template_test.go]: also once again make these outputs generic, prod/stage is fine, but `openshift-monitoring/rick/opf-ci-pipelines/etc` are operate-first specific.

File: alerts/template_test.go [hunk 7]
```diff
@@ -46,34 +57,83 @@ func TestFormatTitleSimple(t *testing.T) {
...
+		t.Run(testNumber, func(t *testing.T) {
+			rh, err := NewReceiver(&fakeClient{}, "default", false, "", nil, tc.titleTmpl, labelsTmplList)
 			if err != nil {
 				t.Fatal(err)
 			}
-
 			title, err := rh.formatTitle(&msg)
+			fmt.Println("title: ", title)
+			labels, err := rh.formatLabels(&msg)
+			fmt.Println("labels: ", labels)
```
Comment (HumairAK) [inline alerts/template_test.go]: don't need print statements

File: alerts/template_test.go [hunk 8]
```diff
@@ -46,34 +57,83 @@ func TestFormatTitleSimple(t *testing.T) {
...
 	}
-
 	for testNum, tc := range tests {
-		testName := fmt.Sprintf("tc=%d", testNum)
-		t.Run(testName, func(t *testing.T) {
-			rh, err := NewReceiver(&fakeClient{}, "default", false, "", nil, tc.tmplTxt)
+		var labelsTmplList []string
+		testNumber := fmt.Sprintf("tc=%d", testNum)
+		testName := fmt.Sprintf("tn=%s", tc.testName)
+		fmt.Println(testName)
```
Comment (HumairAK) [inline alerts/template_test.go]: don't need print statements

File: alerts/template_test.go [hunk 9]
```diff
@@ -93,3 +153,115 @@ func TestFormatTitleSimple(t *testing.T) {
...
+			expectErrTxt: " Value does not exist.", 
+			expectOutput: expectOutputTest6, 
+			outputDescription: "fails, no \"Foo\" key.",
+		},
+	}
+	for testNum, tc := range tests {
+		var labelsTmplList []string
+		testNumber := fmt.Sprintf("tc=%d", testNum)
+		testName := fmt.Sprintf("tn=%s", tc.testName)
+		fmt.Println(testName)
```
Comment (HumairAK) [inline alerts/template_test.go]: don't need print statements

File: alerts/template_test.go [hunk 10]
```diff
@@ -93,3 +153,115 @@ func TestFormatTitleSimple(t *testing.T) {
...
+		labelsTmplList = make([]string, 1)
+		labelsTmplList[0] = tc.labelsTmpl
+		t.Run(testNumber, func(t *testing.T) {
+			rh, err := NewReceiver(&fakeClient{}, "default", false, "", nil, tc.testName, labelsTmplList)
+			if err != nil {
+				t.Fatal(err)
+			}
+			title, err := rh.formatTitle(&msg)
+			fmt.Println("title: ", title)
```
Comment (HumairAK) [inline alerts/template_test.go]: don't need print statements

File: README.md [hunk 11]
```diff
@@ -118,3 +145,13 @@ If the alert includes a `repo` label, issues will be created in that repository,
 under the GitHub organization specified by `-org`. If no `repo` label is
 present, issues will be created in the repository specified by the `-repo`
 option.
+
+## Additional Configuration for Github Spefic Labels
```
Comment (HumairAK) [inline README.md]: I would change 

> Additional Configuration for Github Spefic Labels 

To

> Additional Configuration for Converting alert labels to GitHub labels

File: README.md [hunk 12]
```diff
@@ -118,3 +145,13 @@ If the alert includes a `repo` label, issues will be created in that repository,
...
+
+In adition to mounting the labelmap configuration file, make sure to add the -label-template-file flag as a parameter to the Dockerfile as it is currently.
```
Comment (HumairAK) [inline README.md]: This isn't specific to dockerfile but the app itself, just say something like: 

> To add the label map file mentioned [here](link to above) just you can use the `-label-template-file`. For example: 
```
$alertmanager-github-receiver -authtoken=${token} -org=${org} -repo=${repo} -label-template-file=${path-to-labelmap}
```

**Review by HumairAK — CHANGES_REQUESTED** [review]

**Comment by durandom** [comment]
@Gregory-Pereira is this basically working? If so, I suggest to open a PR already to upstream and get some feedback there.
This is great work and should be exposed 👏

**Comment by Gregory-Pereira** [comment]
excellent idea @durandom. Going to re-review myself and breathe some life into it, before submitting it as an upstream contribution.

**Comment by Gregory-Pereira** [comment]
bump @Gregory-Pereira

**Pull request closed without merging**